go-controller/pkg/factory/factory.go

```diff
@@ -329,28 +329,52 @@ func getObjectMeta(objType reflect.Type, obj interface{}) (*metav1.ObjectMeta, e
 	return nil, fmt.Errorf("cannot get ObjectMeta from type %v", objType)
 }
 
-func (wf *WatchFactory) addHandler(objType reflect.Type, namespace string, sel labels.Selector, funcs cache.ResourceEventHandler, processExisting func([]interface{})) *Handler {
+func getServiceSpec(objType reflect.Type, obj interface{}) (*kapi.ServiceSpec, error) {
+
+	if service, ok := obj.(*kapi.Service); ok {
+		return &service.Spec, nil
+	}
+
+	return nil, fmt.Errorf("cannot get Spec from type %v", objType)
+}
+
+func (wf *WatchFactory) addHandler(objType reflect.Type, namespace string, sel labels.Selector, matchSel labels.Selector, funcs cache.ResourceEventHandler, processExisting func([]interface{})) *Handler {
```

Comment:
addHandler is agnostic to the objectType, with this change we are making it aware of Service types, introducing an argument that is only valid for Services ... breaking the abstraction :smile: